### PR TITLE
Removed Support for Python2

### PR DESCRIPTION
**Pull request by ashah-splunk**
- removed code specific for python version2
- removed six.py dependency which bridged support for python versions 2 and 3
- refactored code as per pylint suggestions
- used python3 features to improve both efficiency and performance
- removed future imports required to support py2 and py3 versions
- updated test cases to use py3 code
- updated CI test configuration to execute test cases in python 3.7 and 3.9 versions
- updated code to remove code deprecation warnings

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ jobs:
       matrix:
         os:
           - ubuntu-latest
-        python: [ 2.7, 3.7 ]
+        python: [ 3.10.7, 3.11.2]
```
Comment (shilpabijam) [inline .github/workflows/test.yml:14]: If I recall we discussed adding versions of python that currently ship with Enterprise?

**Comment by shilpabijam** [comment]
There are some test failures in latest and 8.1 runs, could we address them?

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,18 @@
...
+### Feature updates
+* `ensure_binary`, `ensure_str`, `ensure_text` and `assert_regex` utility methods have been migrated from `six.py` to `splunklib/__init__.py`
+
+### Major changes
+* Removed Code specific to Python2
+* Removed six.py dependency
+* Removed `__future__` imports
+* Refactored & Updated `splunklib` and `tests` to utilise Python3 features
+* Updated CI test matrix to run with Python versions - 3.7, 3.9 and 3.10.7
+* Refactored Code throwing `deprecation` warnings
```
Comment (shilpabijam) [inline CHANGELOG.md:14]: Seeing some deprecation warnings in the test runs, are these expected?

File: splunklib/__init__.py [hunk 3]
```diff
@@ -31,5 +29,50 @@ def setup_logging(level, log_format=DEFAULT_LOG_FORMAT, date_format=DEFAULT_DATE
                         format=log_format,
                         datefmt=date_format)
 
-__version_info__ = (1, 7, 4)
+
+def ensure_binary(s, encoding='utf-8', errors='strict'):
```
Comment (shilpabijam) [inline splunklib/__init__.py:33]: These seem like utility methods that can reside in the utils package?

Comment (shilpabijam) [inline splunklib/__init__.py:33]: I would think a better place to put these utility methods would be the utils package .

File: splunklib/searchcommands/internals.py [hunk 4]
```diff
@@ -54,7 +51,7 @@ def set_binary_mode(fh):
     if sys.version_info >= (3, 0) and hasattr(fh, 'buffer'):
         return fh.buffer
     # check for python3
-    elif sys.version_info >= (3, 0):
+    if sys.version_info >= (3, 0):
```
Comment (jervin-splunk) [inline splunklib/searchcommands/internals.py:54]: Since we are removing support for Python 2 entirely in this MR, this method needs more simplification to remove the checks for sys.version_info - these are no longer necessary.

File: tests/test_job.py [hunk 5]
```diff
@@ -14,23 +14,17 @@
...
-except ImportError:
-    import unittest
+import unittest
 
-import splunklib.client as client
-import splunklib.results as results
+from splunklib import client
+from splunklib import results
 
 from splunklib.binding import _log_duration, HTTPError
```
Comment (jervin-splunk) [inline tests/test_job.py:30]: There's code on line 33 below which refers to Python 2.6 - recommend determining whether the comment is misleading and can now be removed, or whether more work is needed to catch ParseErrors as well.

**Review by jervin-splunk — CHANGES_REQUESTED** [review]
In addition to the places noted in the other comments, there are some lingering references to Python 2.7 and 2.6 which should likely be cleaned up as well in these files:

```
splunk-sdk-python/tests/README.md
splunk-sdk-python/scripts/test_specific.sh
splunk-sdk-python/README.md
```

File: Makefile [hunk 6]
```diff
@@ -34,7 +34,7 @@ docs:
 .PHONY: test
 test:
 	@echo "$(ATTN_COLOR)==> test $(NO_COLOR)"
-	@tox -e py27,py37
+	@tox -e py37,py39
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply Makefile:37]: this was not intentional, it was missed from our side

File: splunklib/searchcommands/external_search_command.py [hunk 7]
```diff
@@ -14,34 +14,31 @@
...
 
-class ExternalSearchCommand(object):
-    """
-    """
+class ExternalSearchCommand:
     def __init__(self, path, argv=None, environ=None):
 
-        if not isinstance(path, (bytes, six.text_type)):
-            raise ValueError('Expected a string value for path, not {}'.format(repr(path)))
+        if not isinstance(path, (bytes,str)):
```
Comment (maszyk99) [inline splunklib/searchcommands/external_search_command.py:37]:
```suggestion
        if not isinstance(path, (bytes, str)):
```

File: .github/workflows/test.yml (hunk 1 above)
Comment (ashah-splunk) [reply .github/workflows/test.yml]: as per the discussion, python 3.7 and Python 3.9 will be part of the matrix.

File: CHANGELOG.md (hunk 2 above)
Comment (ashah-splunk) [reply CHANGELOG.md:14]: yes those warnings are expected as they are regarding couple of features that we have replaced but have to decide on when to remove them permanently from the code.

File: tests/searchcommands/chunked_data_stream.py [hunk 8]
```diff
@@ -4,33 +4,32 @@
...
 
 import splunklib.searchcommands.internals
-from splunklib import six
 
 
-class Chunk(object):
+class Chunk:
     def __init__(self, version, meta, data):
-        self.version = six.ensure_str(version)
+        self.version = version
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply tests/searchcommands/chunked_data_stream.py]: added the method here

File: tests/searchcommands/chunked_data_stream.py [hunk 9]
```diff
@@ -54,7 +53,7 @@ def read_chunk(self):
 
 
 def build_chunk(keyval, data=None):
-    metadata = six.ensure_binary(json.dumps(keyval), 'utf-8')
+    metadata = json.dumps(keyval).encode('utf-8')
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply tests/searchcommands/chunked_data_stream.py]: yes, we missed this. Used `ensure_binary()` method here

File: tests/test_collection.py [hunk 10]
```diff
@@ -174,11 +163,10 @@ def test_list_with_sort_mode_alpha_case(self):
                 kwargs['sort_key'] = 'sid'
             found = [ent.name for ent in coll.list(**kwargs)]
             if len(found) == 0:
-                logging.debug("No entities in collection %s; skipping test.", coll_name)
+                logging.debug(f"No entities in collection {coll_name}; skipping test.")
             expected = sorted(found)
             self.assertEqual(expected, found,
-                             msg='on %s (expected: %s, found: %s)' % \
-                                 (coll_name, expected, found))
+                             msg=f'on {coll_name} (expected {expected}, found {found})')
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply tests/test_collection.py]: rectified the missing colon

File: tests/test_collection.py [hunk 11]
```diff
@@ -154,15 +143,15 @@ def test_list_with_sort_mode_auto(self):
...
+                logging.debug(f"No entities in collection {coll_name}; skipping test.")
 
             if coll_name == 'jobs':
                 found = [ent.name for ent in coll.list(
                     sort_dir="asc", sort_key="sid")]
             else:
                 found = [ent.name for ent in coll.list()]
 
-            self.assertEqual(expected, found, msg='on %s (expected: %s, found: %s)' % (coll_name, expected, found))
+            self.assertEqual(expected, found, msg=f'on {coll_name} (expected {expected}, found {found})')
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply tests/test_collection.py]: rectified the missing colons

File: tests/test_collection.py [hunk 12]
```diff
@@ -193,57 +181,53 @@ def test_list_with_sort_mode_alpha(self):
...
                 offset += count
                 self.assertTrue(count == page_size or offset == total,
-                                msg='on %s' % coll_name)
+                                msg=f'on {coll_name}')
                 found.extend([ent.name for ent in page])
                 logging.debug("Iterate: offset=%d/%d", offset, total)
             self.assertEqual(expected, found,
-                             msg='on %s (expected: %s, found: %s)' % \
-                                 (coll_name, expected, found))
+                             msg=f'on {coll_name} (expected {expected}, found {found})')
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply tests/test_collection.py]: rectified the missing colons.

File: tests/searchcommands/chunked_data_stream.py [hunk 13]
```diff
@@ -87,14 +86,14 @@ def _build_data_csv(data):
...
     for datum in data:
-        headers.update(datum.keys())
+        headers.update(list(datum.keys()))
     writer = csv.DictWriter(csvout, headers,
                             dialect=splunklib.searchcommands.internals.CsvDialect)
     writer.writeheader()
     for datum in data:
         writer.writerow(datum)
-    return six.ensure_binary(csvout.getvalue())
+    return csvout.getvalue().encode('utf-8')
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply tests/searchcommands/chunked_data_stream.py]: yes missed the individual encodes in the code, added `ensure_binary()` method here

File: splunklib/__init__.py (hunk 3 above)
Comment (ashah-splunk) [reply splunklib/__init__.py]: methods have been moved to utils

File: CHANGELOG.md [hunk 14]
```diff
@@ -1,5 +1,18 @@
...
+
+### Feature updates
+* `ensure_binary`, `ensure_str`, `ensure_text` and `assert_regex` utility methods have been migrated from `six.py` to `splunklib/__init__.py`
+
+### Major changes
+* Removed Code specific to Python2
+* Removed six.py dependency
+* Removed `__future__` imports
+* Refactored & Updated `splunklib` and `tests` to utilise Python3 features
+* Updated CI test matrix to run with Python versions - 3.7, 3.9 and 3.10.7
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply CHANGELOG.md]: yes, only 3.7 and 3.9

File: splunklib/client.py [hunk 15]
```diff
@@ -3758,9 +3751,7 @@ def update_accelerated_field(self, name, value):
         :return: Result of POST request
         """
         kwargs = {}
-        if isinstance(value, dict):
-            value = json.dumps(value)
-        kwargs['accelerated_fields.' + name] = value
+        kwargs['accelerated_fields.' + name] = value if isinstance(value, str) else json.dumps(value)
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply splunklib/client.py]: modified the code as per the feedback

File: tests/test_job.py (hunk 5 above)
Comment (ashah-splunk) [reply tests/test_job.py:30]: we had a look and it seems safe to just remove the comments.

File: splunklib/searchcommands/internals.py (hunk 4 above)
Comment (ashah-splunk) [reply splunklib/searchcommands/internals.py]: yes, simplified the method and removed the unnecessary code

File: tests/searchcommands/test_internals_v2.py [hunk 16]
```diff
@@ -418,15 +401,17 @@ def _run(self):
...
 
-        return
-
 
 # test = Test(['random_bytes', 'random_unicode'], [random_bytes, random_unicode])
 # test.record()
 # test.playback()
 
+Test.__test__ = False
+TestRecorder.__test__ = False
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply tests/searchcommands/test_internals_v2.py]: these change may have been added for some testing purpose, not recollecting the reason for adding these.
Have removed this change after analysing if they have any impact on test execution.

File: tests/searchcommands/test_decorators.py [hunk 17]
```diff
@@ -474,11 +457,12 @@ def test_option(self):
...
             'required_set="bar" set="bar" show_configuration="f"')
 
-        observed = six.text_type(command.options)
+        observed = str(command.options)
 
         self.assertEqual(observed, expected)
-        return
 
 
+TestSearchCommand.__test__ = False
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply tests/searchcommands/test_decorators.py]: this change may have been added for some testing purpose, not recollecting the reason for adding these.
Have removed this change after analysing if they have any impact on test execution.

File: tests/searchcommands/chunked_data_stream.py [hunk 18]
```diff
@@ -4,33 +4,32 @@
...
+class ChunkedDataStreamIter(collections.abc.Iterator):
     def __init__(self, chunk_stream):
         self.chunk_stream = chunk_stream
 
     def __next__(self):
-        return self.next()
+        return next(self)
 
-    def next(self):
+    def __next__(self):
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply tests/searchcommands/chunked_data_stream.py]: rectified the mistake.

File: splunklib/searchcommands/search_command.py [hunk 19]
```diff
@@ -978,25 +965,25 @@ def _execute_v2(self, ifile, process):
             metadata, body = result
             action = getattr(metadata, 'action', None)
             if action != 'execute':
-                raise RuntimeError('Expected execute action, not {}'.format(action))
+                raise RuntimeError(f'Expected execute action, not {action}')
 
             self._finished = getattr(metadata, 'finished', False)
             self._record_writer.is_flushed = False
-
+            self._metadata.update(metadata)
```
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply splunklib/searchcommands/search_command.py:972]: To add access for finished flag within metadata, update method was added 
code commit Ref - https://github.com/splunk/splunk-sdk-python/pull/456/commits/bf40010ed603c1f35a72ea9b437f0c672024c8e3

File: splunklib/searchcommands/external_search_command.py (hunk 7 above)
Comment (ashah-splunk) [reply splunklib/searchcommands/external_search_command.py:37]: will check for formatting in each file and commit the changes together

File: splunklib/searchcommands/internals.py [hunk 20]
```diff
@@ -687,7 +669,6 @@ def _write_record(self, record):
         # We may be running under PyPy 2.5 which does not include the _json module
         _iterencode_json = JSONEncoder(separators=(',', ':')).iterencode
     else:
-        # Creating _iterencode_json this way yields a two-fold performance improvement on Python 2.7.9 and 2.7.10
```
Comment (jervin-splunk) [inline splunklib/searchcommands/internals.py:690]: If we are not cleaning up the code which purports to support PyPy 2.5 - which we *definitely* do not support at this stage given that that version was released in 2015 - then we should reinstate this comment, otherwise the code does not make sense.

Ideally, this entire block would be refactored to  modern standards; importing the private implementation of the module by using the underscore name should really no longer be necessary.

**Review by ichaer — CHANGES_REQUESTED** [review]

File: splunklib/binding.py [hunk 21]
```diff
@@ -121,8 +119,8 @@ def _parse_cookies(cookie_str, dictionary):
     :param dictionary: A dictionary to update with any found key-value pairs.
     :type dictionary: ``dict``
     """
-    parsed_cookie = six.moves.http_cookies.SimpleCookie(cookie_str)
-    for cookie in parsed_cookie.values():
+    parsed_cookie = SimpleCookie(cookie_str)
+    for cookie in list(parsed_cookie.values()):
```
[the remark this answers is not in the document]
Comment (maszyk99) [reply splunklib/binding.py:123]: I fixed it for files that were edited in this PR. I will create a task to remove it from all other places.

File: splunklib/client.py [hunk 22]
```diff
@@ -2360,22 +2342,21 @@ def __getitem__(self, key):
...
-                        raise AmbiguousReferenceException("Found multiple inputs of kind %s named %s." % (kind, key))
-                    elif len(entries) == 0:
+                        raise AmbiguousReferenceException(f"Found multiple inputs of kind {kind} named {key}.")
+                    if len(entries) == 0:
                         pass
-                    else:
-                        if candidate is not None: # Already found at least one candidate
-                            raise AmbiguousReferenceException("Found multiple inputs named %s, please specify a kind" % key)
-                        candidate = entries[0]
+                    if candidate is not None:  # Already found at least one candidate
```
[the remark this answers is not in the document]
Comment (maszyk99) [reply splunklib/client.py:2348]: @ashah-splunk We will need your assitance here

File: splunklib/searchcommands/internals.py [hunk 23]
```diff
@@ -799,30 +780,31 @@ def write_chunk(self, finished=None):
...
 
     def write_metadata(self, configuration):
         self._ensure_validity()
 
-        metadata = chain(six.iteritems(configuration), (('inspector', self._inspector if self._inspector else None),))
+        metadata = chain(configuration.items(), (('inspector', self._inspector if self._inspector else None),))
         self._write_chunk(metadata, '')
-        self.write('\n')
+        # Removed additional new line 
+        # self.write('\n')
```
[the remark this answers is not in the document]
Comment (maszyk99) [reply splunklib/searchcommands/internals.py:793]: @ashah-splunk

File: splunklib/client.py [hunk 24]
```diff
@@ -3483,14 +3465,16 @@ def update(self, **kwargs):
...
     @property
     def role_entities(self):
         """Returns a list of roles assigned to this user.
 
         :return: The list of roles.
         :rtype: ``list``
         """
-        return [self.service.roles[name] for name in self.content.roles]
+        all_role_names = [r.name for r in self.service.roles.list()]
+        return [self.service.roles[name] for name in self.content.roles if name in all_role_names]
```
[the remark this answers is not in the document]
Comment (maszyk99) [reply splunklib/client.py:3476]: Ideally we should split it into another pull request and review it just after this one gets merged, but maybe this change is needed as fast as possible. What do you think @ashah-splunk ?

[the remark this answers is not in the document]
Comment (ashah-splunk) [reply splunklib/client.py:3476]: Hi @ichaer,
this change was made in reference to the issue https://github.com/splunk/splunk-sdk-python/issues/527.
As per the discussions with the team then, we decided to add this as a part of this PR. 
We have tested the code as well as its been reviewed by the team then before pushed to the PR.

File: splunklib/searchcommands/search_command.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply splunklib/searchcommands/search_command.py:972]: Hi @ichaer ,
This change was made in reference to the slack thread - https://splunk.slack.com/archives/CA9R2GU4U/p1669930999509219
It was reported by some customer so it was considered to be added in this PR. At that time plan was in-motion to release the Python 2.0.0-beta version, but then the Org restructure happened and the release was postponed.

File: splunklib/searchcommands/internals.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply splunklib/searchcommands/internals.py:793]: Hi @ichaer 
To my knowledge this change was made in reference to the Jira ticket - https://splunk.atlassian.net/browse/DVPL-11850.
As per the comments on the ticket and response by Navya - https://splunk.atlassian.net/browse/DVPL-11850?focusedCommentId=10596736
This was at that time planned as part of the refactor hence added to this PR.

File: splunklib/client.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (ichaer) [reply splunklib/client.py:3476]: OK, I won't block you on this, but I would urge you to review your change process for future PRs. PRs are cheap to create and merge, there's no reason to group unrelated changes into one PR, and splitting them allows for more focused reviews and testing, as well as easier reverts.

File: splunklib/searchcommands/internals.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (ichaer) [reply splunklib/searchcommands/internals.py:793]: OK, if we want to keep the change let's remove the commented-out code. History is already preserved by git, let's leave it out of the code.

File: splunklib/client.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply splunklib/client.py]: sure @ichaer , we will make sure to have separate PRs going further. Thanks!

File: splunklib/searchcommands/internals.py (hunk 23 above)
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply splunklib/searchcommands/internals.py]: Ok, we have removed the commented out code.

File: splunklib/client.py (hunk 22 above)
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply splunklib/client.py]: @ichaer we have added the `else` block back.

File: splunklib/searchcommands/internals.py (hunk 20 above)
Comment (maszyk99) [reply splunklib/searchcommands/internals.py]: I tried to refactor entire block to modern standards by simply adding function with `json.dumps`, but there are some corner cases that this method doesn't cover and it causes one test to fail.

For the sake of the release, I just added the comment back for now.

Comment (jervin-splunk) [reply splunklib/searchcommands/internals.py]: Sounds good.

File: splunklib/searchcommands/external_search_command.py [hunk 25]
```diff
@@ -120,13 +117,13 @@ def _execute(path, argv=None, environ=None):
             found = ExternalSearchCommand._search_path(path, search_path)
 
             if found is None:
-                raise ValueError('Cannot find command on path: {}'.format(path))
+                raise ValueError(f'Cannot find command on path: {path}')
 
             path = found
-            logger.debug('starting command="%s", arguments=%s', path, argv)
+            logger.debug(f'starting command="{path}", arguments={path}')
```
[the remark this answers is not in the document]
Comment (ichaer) [reply splunklib/searchcommands/external_search_command.py]: You haven't acted on this, the log line is still wrong.

File: splunklib/searchcommands/search_command.py [hunk 26]
```diff
@@ -738,29 +725,29 @@ def _process_protocol_v2(self, argv, ifile, ofile):
...
+                            self.write_error(f'Illegal value: {name}={value}')
                             error_count += 1
                             continue
 
             missing = self.options.get_missing()
 
             if missing is not None:
                 if len(missing) == 1:
-                    self.write_error('A value for "{}" is required'.format(missing[0]))
+                    self.write_error(f'A value for "{missing[0]}" is required'.format())
```
[the remark this answers is not in the document]
Comment (ichaer) [reply splunklib/searchcommands/search_command.py]: This hasn't been addressed either.

**Comment by ichaer** [comment]
I'm approving anticipating that the two pending comments will be resolved. They are not about particularly concerning issues, but they are trivial to implement and shouldn't be postponed.

**Review by ichaer — APPROVED** [review]

File: splunklib/searchcommands/search_command.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply splunklib/searchcommands/search_command.py]: @ichaer this change has been applied, I think we missed to update here.
commit ref - https://github.com/splunk/splunk-sdk-python/pull/456/commits/9053482a3ef97e9d04684fec9d4117f1e40f6394

File: splunklib/searchcommands/external_search_command.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (ashah-splunk) [reply splunklib/searchcommands/external_search_command.py]: @ichaer we have resolved this as well. please find the commit here - https://github.com/splunk/splunk-sdk-python/pull/456/commits/1d1cc207f131682e02773f54df80960def228245.
sorry we missed to update here.
